bee-protocol/src/worker/mod.rs

```diff
@@ -40,3 +40,17 @@ pub(crate) use sender::SenderWorker;
 pub(crate) use status::StatusWorker;
 pub(crate) use tps::TpsWorker;
 pub(crate) use transaction::{HasherWorker, HasherWorkerEvent, ProcessorWorker};
+
+use bee_common::worker::Error as WorkerError;
+
+use async_trait::async_trait;
+use futures::Stream;
+
+#[async_trait]
+pub(crate) trait Worker {
+    type Event;
+
+    type Receiver: Stream<Item = Self::Event>;
+
+    async fn run(self, receiver: Self::Receiver) -> Result<(), WorkerError>;
```

Comment:
I wonder whether we could separate `run` into distinct `start` and `stop` methods such that we can know externally when each worker has started and when it has finished its shutdown (right now it's a bit opaque and this is going to be annoying to work with when we want more granular control over startup/shutdown)

Reply:
@zesterer what should `stop` do? I think most (all?) workers just break the loop and write the `Stopped` log entry.

Reply:
Yep, `stop` would break the loop. I do wonder whether the existing shutdown system might have a part to play here instead though.